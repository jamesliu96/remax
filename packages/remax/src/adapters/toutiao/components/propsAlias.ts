import kebabCase from 'lodash.kebabcase';
import plainStyle from '../../../utils/plainStyle';

const alias: { [prop: string]: string } = {};

export function getAlias(prop: string, isNative = false) {
  const aliasProp = alias[prop];

  if (aliasProp) {
    return aliasProp;
  }

  if (prop.endsWith('className')) {
    return prop.replace('className', 'class');
  }

  if (prop.startsWith('on')) {
    prop = prop.toLowerCase().replace('on', 'bind');

    if (!isNative) {
      prop = prop.replace('click', 'tap');
    }

    return prop;
  }

  return kebabCase(prop);
}

function getValue(prop: string, value: any): any {
  if (prop.toLowerCase().endsWith('style') && prop !== 'layerStyle') {
    return plainStyle(value);
  }

  return value;
}

export interface GenericProps {
  [key: string]: any;
}

export default function propsAlias(props: GenericProps, isNative = false) {
  const aliasProps: GenericProps = {};

  Object.keys(props).forEach(prop => {
    aliasProps[getAlias(prop, isNative)] = getValue(prop, props[prop]);
  });

  return aliasProps;
}
