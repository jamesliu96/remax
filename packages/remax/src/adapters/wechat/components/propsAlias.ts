import kebabCase from 'lodash.kebabcase';
import plainStyle from '../../../utils/plainStyle';

const alias: { [prop: string]: string } = {
  activeColor: 'activeColor',
  backgroundColor: 'backgroundColor',
  onClick: 'bindtap',
  catchClick: 'catchtap',
  enable3D: 'enable-3D',
  hTouchMove: 'htouchmove',
  vTouchMove: 'vtouchmove',
};

export function getAlias(prop: string, isNative = false) {
  const aliasProp = alias[prop];

  if (aliasProp) {
    return aliasProp;
  }

  if (prop.endsWith('className')) {
    prop = prop.replace('className', 'class');
  }

  if (prop.startsWith('on') || prop.startsWith('catch')) {
    prop = prop.toLowerCase().replace('on', 'bind');

    if (!isNative) {
      prop = prop.replace('click', 'tap');
    }

    return prop;
  }

  return kebabCase(prop);
}

function getValue(prop: string, value: any): any {
  if (prop.toLowerCase().endsWith('style') && prop !== 'layerStyle') {
    return plainStyle(value);
  }

  return value;
}

export interface GenericProps {
  [key: string]: any;
}

export default function propsAlias(props: GenericProps, isNative = false) {
  const aliasProps: GenericProps = {};

  Object.keys(props).forEach(prop => {
    aliasProps[getAlias(prop, isNative)] = getValue(prop, props[prop]);
  });

  return aliasProps;
}
