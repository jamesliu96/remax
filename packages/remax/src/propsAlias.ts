import kebabCase from 'lodash.kebabcase';
import Platform from './Platform';
import plainStyle from './utils/plainStyle';

function functionPropAlias(prop: string, isNative = false, platform?: string) {
  if (!isNative) {
    prop = prop.replace('Click', 'Tap');
  }

  if (platform === 'alipay') {
    return prop;
  }

  prop = prop.toLowerCase().replace('on', 'bind');

  return prop;
}

export function getAlias(prop: string, isNative = false, platform?: string) {
  prop = prop.replace('className', 'class').replace('ClassName', 'Class');

  if (prop.startsWith('on') || prop.startsWith('catch')) {
    return functionPropAlias(prop, isNative, platform);
  }

  return kebabCase(prop);
}

function getValue(prop: string, value: any): any {
  if (prop.toLowerCase().endsWith('style') && prop !== 'layerStyle') {
    return plainStyle(value);
  }

  return value;
}

export interface GenericProps {
  [key: string]: any;
}

export default function propsAlias(
  props: GenericProps,
  isNative = false,
  platform = Platform.current
) {
  if (!props) {
    return props;
  }

  const aliasProps: GenericProps = {};

  Object.keys(props).forEach(prop => {
    aliasProps[getAlias(prop, isNative, platform)] = getValue(
      prop,
      props[prop]
    );
  });

  return aliasProps;
}
