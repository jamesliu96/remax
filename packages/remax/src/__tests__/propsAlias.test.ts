import propsAlias, { getAlias } from '../propsAlias';

describe('propsAlias', () => {
  it('transform className prop correctly', () => {
    expect(getAlias('className')).toBe('class');
    expect(getAlias('placeholderClassName')).toBe('placeholder-class');
  });

  it('transform function prop correctly', () => {
    expect(getAlias('onTouchStart')).toBe('bindtouchstart');
    expect(getAlias('onClick')).toBe('bindtap');
    expect(getAlias('catchClick')).toBe('catchtap');
    expect(getAlias('onLongClick')).toBe('bindlongtap');
    expect(getAlias('onClick', true)).toBe('bindclick');

    expect(getAlias('onTouchStart', false, 'alipay')).toBe('onTouchStart');
    expect(getAlias('onClick', false, 'alipay')).toBe('onTap');
    expect(getAlias('catchClick', false, 'alipay')).toBe('catchTap');
    expect(getAlias('onLongClick', false, 'alipay')).toBe('onLongTap');
    expect(getAlias('onClick', true, 'alipay')).toBe('onClick');
  });

  it('transform style prop correctly', () => {
    expect(
      propsAlias({
        style: {
          color: '#ffffff',
        },
      })
    ).toMatchSnapshot();
  });

  afterEach(() => {
    process.env.REMAX_PLATFORM = undefined;
  });
});
