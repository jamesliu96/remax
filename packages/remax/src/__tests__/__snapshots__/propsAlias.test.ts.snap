// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`propsAlias transform style prop correctly 1`] = `
Object {
  "style": "color:#ffffff;",
}
`;
